diff-containers/test/Test/Data/Map/Diff/Strict.hs

```diff
@@ -1,66 +1,100 @@
 
-import           Data.Semigroupoid.Simple.Auto
 import           Data.Semigroupoid.Simple.Laws
 
+import           Test.Util
+
 tests :: TestTree
 tests = testGroup "Data.Map.Diff.Strict" [
-      testGroupWithProxy (Proxy @(DiffEntry (Smaller Int))) [
-      ]
-    , testGroupWithProxy (Proxy @(DiffHistory (Smaller Int))) [
+      localOption (QuickCheckTests 1000) $
+      testGroupWithProxy (Proxy @(DiffHistory (OftenSmall Int))) [
           testSemigroupLaws
         , testMonoidLaws
         , testGroupLaws
         ]
-    , testGroupWithProxy (Proxy @(Auto (DiffHistory (Smaller Int)))) [
-          testSemigroupoidLaws
-        , testGroupoidLaws
-        ]
-    , testGroupWithProxy (Proxy @(Diff (Smaller Int) (Smaller Int))) [
+    , localOption (QuickCheckTests 1000) $
+      testGroupWithProxy (Proxy @(Diff (OftenSmall Int) (OftenSmall Int))) [
           testSemigroupLaws
         , testMonoidLaws
         , testGroupLaws
         ]
-    , testGroupWithProxy (Proxy @(Auto (Diff (Smaller Int) (Smaller Int)))) [
-          testSemigroupoidLaws
-        , testGroupoidLaws
-        ]
-    , testProperty "prop_diffThenApply @(Smaller Int)" $
-        prop_diffThenApply @(Smaller Int) @(Smaller Int)
-    , testProperty "prop_diffThenApply @Int" $
-        prop_diffThenApply @Int @Int
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_diffingIsPositive" $
+        prop_diffingIsPositive @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_diffThenApply" $
+        prop_diffThenApply @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_applyMempty" $
+        prop_applyMempty @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckMaxRatio 100) $
+      localOption (QuickCheckTests 1000) $
+      testProperty "prop_applyAllAndApplySum" $
+        prop_applyAllAndApplySum @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckMaxRatio 100) $
+      localOption (QuickCheckTests 1000) $
+      testProperty "prop_unsafeApplyAllAndUnsafeApplySum" $
+        prop_unsafeApplyAllAndUnsafeApplySum @(OftenSmall Int) @(OftenSmall Int)
     ]
 
 {------------------------------------------------------------------------------
   Simple properties
 ------------------------------------------------------------------------------}
 
+prop_diffingIsPositive ::
+     (Ord k, Eq v)
+  => Map k v
+  -> Map k v
+  -> Property
+prop_diffingIsPositive m1 m2 = property $ isPositive (diff m1 m2)
+
 prop_diffThenApply ::
      (Show k, Show v, Ord k, Eq v)
   => Map k v
   -> Map k v
   -> Property
-prop_diffThenApply m1 m2 = applyDiff m1 (diff m1 m2) === m2
+prop_diffThenApply m1 m2 = applyDiff m1 (diff m1 m2) === Right m2
+
+prop_applyMempty ::
+     (Show k, Show v, Ord k, Eq v)
+  => Map k v
+  -> Property
+prop_applyMempty m = applyDiff m mempty === Right m
+
+prop_applyAllAndApplySum ::
```

Comment:
Maybe we can add a comment on why we require all the diffs to be positive. What would happen otherwise?

Reply:
A negative diff can't be applied to a `Data.Map`. If, for example, we would have two diffs and their sum
* `d1 = [1 -> [UnsafeAntiInsert 1]]` (`d1` maps a key `1` to a singleton diff history) 
* `d2 = [1 -> [Insert 1]]` (`d1` maps a key `1` to a singleton diff history)
* `d3 = d1 <> d2 = []` (because `UnsafeAntiInsert 1` and `Insert 1` cancel out)

Applying `d3` is the identity function. Applying `d1` first and then applying `d2` should also be the identity function, but there's currently no way to represent the result of applying `d1` first such that applying `d2` after will cancel out `d1`.

It is an unfortunate side effect of facilitating a `Group` instance for diffs that we can get into this situation. The responsibility of downstream code is to ensure that diffs that are applied are positive (i.e., no diff histories should include `UnsafeAnti*` diff entries). If that is the case, then applying diffs will never go wrong.

I will comment on this in the code, and I will also dedicate a section of the `diff-containers` README to this "positivity" property (#8).

Reply:
For a somewhat similar reason, diffs and diff histories should also be "normal" (i.e., the diff histories have no consecutive elements that can cancel out). I will comment on this as well.

Reply:
I've decided to comment on the positivity in a separate PR, because as I writing the comments I found that I should be testing some of the statements I make.

Reply:
See #16